core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -164,162 +185,159 @@ class TransactionStateManager(brokerId: Int,
+                Some(TransactionResult.ABORT, txnMetadata.prepareComplete(time.milliseconds()))
+              case PrepareCommit =>
+                Some(TransactionResult.COMMIT, txnMetadata.prepareComplete(time.milliseconds()))
+              case _ =>
+                // nothing need to be done
+                None
+            }
+          }
+
+          result.foreach { case (command, newMetadata) =>
+            sendTxnMarkers(transactionalId, coordinatorEpoch, command, txnMetadata, newMetadata)
+          }
+      }
+
+      inLock(stateLock) {
+        addLoadedTransactionsToCache(topicPartition.partition, coordinatorEpoch, loadedTransactions)
+        loadingPartitions.remove(partitionId)
       }
     }
 
-    scheduler.schedule(topicPartition.toString, loadTransactions _)
+    scheduler.schedule(s"load-txns-for-partition-$topicPartition", loadTransactions _)
   }
 
   /**
    * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
    * that belong to that partition.
    */
-  def removeTransactionsForPartition(partition: Int) {
+  def removeTransactionsForTxnTopicPartition(partitionId: Int) {
     validateTransactionTopicPartitionCountIsStable()
 
-    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
-
-    inLock(stateLock) {
-      ownedPartitions.remove(partition)
-      loadingPartitions.remove(partition)
-    }
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partitionId)
 
     def removeTransactions() {
-      var numTxnsRemoved = 0
-
       inLock(stateLock) {
-        for (transactionalId <- transactionMetadataCache.keys) {
-          if (partitionFor(transactionalId) == partition) {
-            // we do not need to worry about whether the transactional id has any ongoing transaction or not since
-            // the new leader will handle it
-            transactionMetadataCache.remove(transactionalId)
-            numTxnsRemoved += 1
-          }
+        transactionMetadataCache.remove(partitionId) match {
+          case Some(txnMetadataCacheEntry) =>
+            info(s"Removed ${txnMetadataCacheEntry.metadataPerTransactionalId.size} cached transaction metadata for $topicPartition on follower transition")
+
+          case None =>
+            info(s"Trying to remove cached transaction metadata for $topicPartition on follower transition but there is no entries remaining; " +
+              s"it is likely that another process for removing the cached entries has just executed earlier before")
         }
 
-        if (numTxnsRemoved > 0)
-          info(s"Removed $numTxnsRemoved cached transaction metadata for $topicPartition on follower transition")
+        loadingPartitions.remove(partitionId)
       }
     }
 
-    scheduler.schedule(topicPartition.toString, removeTransactions _)
+    scheduler.schedule(s"remove-txns-for-partition-$topicPartition", removeTransactions _)
```

Comment:
Got it. However, if we just do the removal in the handler thread, then the immigrate-then-emigrate issue may occur. On the other hand, we cannot depend on scheduler do not have pending request since we are periodically schedule txn-expiration and pid-expiration with that scheduler as well.

So here is what I will do: add a removal partitions which will be modified in the handler thread, then get txn metadata will check if the txn partition is in the removal partitions set.

Then in the locking PR, I will add the read-write-lock so that checking will be monitored by the read lock. WDYT?